src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,232 @@
+              List.exists (is_rec_typ typ_names) cts
+         | Rinherit ct -> is_rec_typ typ_names ct) rws
   | _ -> false
 
-let name s =
-  let prefix = "gen" in
-  match s with "t" -> prefix | s -> prefix ^ "_" ^ s
+let is_rec_constr_decl typ_names cd =
+  match cd.pcd_args with
+  | Pcstr_tuple cts -> List.exists (is_rec_typ typ_names) cts
+  | _ -> false
 
-let pat ~loc s =
-  let (module A) = Ast_builder.make loc in
-  let s = name s in
-  A.pvar s
+(** [is_rec_type_decl typ_names typ] looks for elements of [typ_names]
+    recursively in [typ]. *)
+let is_rec_type_decl typ_names typ =
+  let in_type_kind =
+    match typ.ptype_kind with
+    | Ptype_variant cstrs -> List.exists (is_rec_constr_decl typ_names) cstrs
+    | _ -> false
+  in
+  let in_manifest =
+    match typ.ptype_manifest with
+    | Some x -> is_rec_typ typ_names x
+    | None -> false
+  in
+  in_type_kind || in_manifest
+
+(** [env] contains the list of recursive types during the derivation
+
+    i.e:
+    - contains one element maximum if its a single type declaration *)
+type env = {
+    rec_types : string list;
+    curr_types : string list;
+    curr_type : string;
+  }
 
-let gen ~loc ?(env = TypeGen.empty) lg =
+(** [add_if_rec env tys ty x] adds x to [env.rec_types] if typ is recursive *)
+let add_if_rec env typ x =
+  if is_rec_type_decl env.curr_types typ then
+    { env with rec_types = x :: env.rec_types}
+  else env
+
+(** {2. Generator constructors} *)
+
+(** [gen_longident lg args] creates a generator using [lg].
+
+    The longident can either be a:
+    - Lident s: We transform to gen_s (or gen if s = "t")
+    - Ldot (lg, s): We transform to qualified generator (e.g. B.gen)
+*)
+let gen_longident ~loc ~env lg args =
   let (module A) = Ast_builder.make loc in
   match lg with
   | Lident s ->
-      Option.value ~default:(name s |> A.evar) @@ TypeGen.find_opt s env
-  | Ldot (lg, s) -> A.(pexp_ident (Located.mk @@ Ldot (lg, name s)))
+     if List.mem s env.rec_types then
+       name ~sized:true s |> A.evar |>
+         Args.apply_args ~loc args |>
+         Args.apply_args ~loc [ [%expr (n / 2)] ]
+     else
+       name s |> A.evar |> Args.apply_args ~loc args
+  | Ldot (lg, s) ->
+     A.(pexp_ident (Located.mk @@ Ldot (lg, name s))) |>
+       Args.apply_args ~loc args
   | Lapply (_, _) -> raise (Invalid_argument "gen received an Lapply")
 
-let frequency ~loc l = [%expr QCheck.Gen.frequency [%e l]]
+(** [gen_sized typ_name is_rec to_gen xs] uses [is_rec] to determine recursive
+    nodes in [xs].
 
-let pure ~loc x = [%expr QCheck.Gen.pure [%e x]]
+    If no recursive node is found, the type is _not_ recursive, we build a
+    generator using frequency.
 
-let tree ~loc nodes leaves =
-  [%expr
-    QCheck.Gen.sized
-    @@ QCheck.Gen.fix (fun self -> function
-         | 0 -> [%e leaves] | n -> [%e nodes])]
+    However, if recursive nodes are found, we build a tree like generator using
+    {!tree}.
```

Comment:
Needs quick update as `tree` is no longer in PR code.

Reply:
Great catch :) https://github.com/c-cube/qcheck/pull/195/commits/0ac00e4c992d9ff41a737eb8ffdda7a598ab9b38